Add convenience layer to FineTuning with tests

mikhail opened the pull request:
Work in this PR:
# OpenAI
- [x] Add tests for FineTuning.
- [x] Add CreateJob with required parameters.
- [x] Add CancelJob.
- [x] Add Async methods.
- [x] Add hyperparameters.
- [x] Add suffix.
- [x] Add validation file.
- [x] Add integrations.
- [x] Add seed.
- [x] Add Events
- [x] Add Checkpoints
- [x] Add Checkpoint-Metrics

# Azure OpenAI
- [x] Delete endpoint
- [x] Tests

mikhail committed:
Add convenience layer to FineTuning with tests

This does not YET include hyperparameters

mikhail committed:
Regenerate API summary

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -34,15 +34,16 @@ public class OpenAIClient {
         public virtual VectorStoreClient GetVectorStoreClient();
     }
-    public class OpenAIClientOptions : ClientPipelineOptions {
+    public class OpenAIClientOptions {
```
The `export-api.ps1` script is behaving a bit unexpectedly; I'm guessing the rigid `--assembly-reference` paths in the script don't match your local environment and it's silently not picking up the dependencies for clean parity output.

Doesn't need to be fixed right now and the exported API is still very useful to see the new surface, but we can (a) see if we can get a clean baseline for you and (b) should see if we can make those assembly references more robust somehow.

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1726,9 +1357,13 @@ public class FineTuningClient {
         public virtual Task<ClientResult> CancelJobAsync(string jobId, RequestOptions options);
+        public Task<ClientResult<FineTuningJob>> CancelJobAsync(string jobId, CancellationToken cancellationToken = default);
         public virtual ClientResult CreateJob(BinaryContent content, RequestOptions options = null);
+        public ClientResult<FineTuningJob> CreateJob(string training_file, string model, RequestOptions options = null);
```
`training_file` -> `trainingFile` or maybe `inputFilename`; see [.NET conventions](https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/coding-style/identifier-names#naming-conventions) for the "method parameters use camelCase" and other idiomatic stipulations. The [Azure SDK .NET guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html) are really handy, too, though we don't necessary match 100% of that given this isn't fundamentally an Azure SDK library.

Also -- and this is minor -- what are your thoughts on the parameter ordering? I'd intuitively think of model before filename ("I want to create a job that fine-tunes gpt-35-turbo using the input from this file here") but the other way isn't illogical, either ("I want to use this input file here to create a fine-tuned version of gpt-35-turbo").

mikhail replied:
Changed the order; renamed to trainingFile. Decided against inputFilename to stay closer to openai's docs and payload.

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+    public class FineTuningJob {
+        public DateTimeOffset CreatedAt { get; }
+        public FineTuningJobError Error { get; }
+        public DateTimeOffset? EstimatedFinish { get; }
```
We should customize the name of this property to be consistent, e.g. `EstimatedFinishAt` (with the 'At' suffix).

mikhail replied:
Agree that it's inconsistent, but this is what OpenAI uses in documentation and payload. I could be persuaded if you feel strongly

trrwilson replied:
One example of established precedent for the library: we renamed `ChatCompletion.Created` to `CreatedAt`.

It's case-by-case consideration and we don't want to make changes willy-nilly, but the first priority is for the library to be usable, idiomatic, and intuitive -- well-reasoned renames and structural deviations are worthy of consideration in that context.

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+        public DateTimeOffset CreatedAt { get; }
+        public FineTuningJobError Error { get; }
+        public DateTimeOffset? EstimatedFinish { get; }
+        public string FineTunedModel { get; }
```
Mostly just a note for later: we'll want to do a holistic naming pass to see what, if any, non-trivial naming deviations we'd like to make from the REST. I think "CreatedModelName" or similar here would be a lot more descriptive and accurate in a .NET API (similarly, something like "baseModelName" on the input) but that has to be weighed against spec/docs divergence and the likely level of existing knowledge/background fine-tuning customers will already have.

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+        public string Model { get; }
+        public string Object { get; }
+        public string OrganizationId { get; }
+        public IReadOnlyList<string> ResultFiles { get; }
```
I think we'll want this to be `ResultFileIds`

mikhail replied:
same as before. But if we're using `ids` here then during job creation it ought to also be file id rather than file name

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+        public FineTuningJobHyperparameters? Hyperparameters { get; set; }
+        public string Id { get; }
+        public string Model { get; }
+        public string Object { get; }
```
These "Object" properties can be hidden (made internal) -- strong types render it redundant. See e.g. `ChatCompletion.cs` for an example of where `Object` is hidden (in that case, made private).

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+        public FineTuningJobStatus Status { get; }
+        public int? TrainedTokens { get; }
+        public string TrainingFile { get; }
+        public string ValidationFile { get; }
```
Like with `ResultFiles`, we'll want to make the minor suffix addition to `TrainingFileId` and `ValidationFileId`.

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+    }
+    public partial struct FineTuningJobHyperparameters {
+        private object _dummy;
+        private int _dummyPrimitive;
```
No idea why these private fields are being emitted -- something's definitely funky the tool!

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+        public string Message { get; }
+        public string Param { get; }
+    }
+    public partial struct FineTuningJobHyperparameters {
```
Reading this closely, it looks like there's a significant difference between the request/input representation of Hyperparameters and the response/output representation -- the request side is a lot more complicated!

I'd suggest:
- Create a new `FineTuningHyperparameterOptions` class that uses `[CodeGenModel("CreateFineTuningJobRequestHyperparameters")] as its basis -- I believe that's the generated source material for the request side (and why things likely appeared to be missing)
- Keep the existing type for the response model (which will still need to solve the "int or string" problem but not have the other fields)
- We can consider (later) if we want the response type to more transparently differentiated by name. e.g. "FineTuningJobHyperparameterInfo" or similar

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1726,9 +1357,13 @@ public class FineTuningClient {
         public virtual Task<ClientResult> CancelJobAsync(string jobId, RequestOptions options);
+        public Task<ClientResult<FineTuningJob>> CancelJobAsync(string jobId, CancellationToken cancellationToken = default);
         public virtual ClientResult CreateJob(BinaryContent content, RequestOptions options = null);
+        public ClientResult<FineTuningJob> CreateJob(string training_file, string model, RequestOptions options = null);
```
The pattern to follow for method parameters:

**Protocol** (should already be present, but here for contrast):
- Operation parameters that *aren't* part of the request body (path, query string, etc.) form the first parameters of the method
- If the operation accepts a request body, it also has a `BinaryContent` parameter
- All protocol methods then end with a `RequestOptions` parameter -- which should be optional if there's a BinaryContent parameter and required if there isn't (this addresses signature overlap)

**Convenience methods**:
- All required parameters, no matter what kind they are (query string, path, request body, etc.) are "promoted" into method parameters
- Optional parameters all go into an "*Options" class (like `ChatCompletionOptions`) that's the second-to-last, optional parameter
- The final parameter is `CancellationToken cancellationToken = default`

Convenience methods then just build off the protocol methods:
- Emplace any parameters that were "promoted" out of the request body back into the Options model instance
- Call the protocol method using the same path/query/etc. parameters and the output of `.ToBinaryContent()` from options for the body -- use `cancellationToken.ToRequestOptions()` for that parameter
- Deserialize the raw response from the protocol result into the desired output model and then use CilentResult.FromValue(...) for the convenience result

Applying that all here, a signature for `CreateJob` would be something like:
```csharp
public virtual ClientResult<FineTuningJob> CreateJob(string baseModelName, string trainingFileId, FineTuningJobOptions options = null, CancellationToken cancellationToken = default)
```

`FineTuningJobOptions` (or `FineTuningJobCreationOptions`) would be from the generated `CreateFineTuningJobRequest` type, just with the "promoted" `model` and `training_file` properties hidden (made internal).

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1738,26 +1373,70 @@ public class FineTuningClient {
+        public FineTuningJobHyperparameters(BinaryData nEpochs);
+        public BinaryData BatchSize { get; }
+        public BinaryData LearningRateMultipler { get; }
+        public BinaryData NEpochs { get; }
```
Each of these is a weird union type that needs massaging and a rename -- I think it sets the record for highest concentration! We can use something like `ChatToolChoice` as a pattern to match (and I can follow up about customizing the code generation input the same way we did for that type -- this doesn't block wiring it up but will improve how much of it can be generated).

We probably want all of these to be possible (making some names up on the fly):
```csharp
options.BatchSizeSelection = HyperparameterBatchSizeSelection.Auto;
options.BatchSizeSelection = new HyperparameterBatchSizeSelection(42);
options.BatchSizeSelection = 42; // via implicit operator
options.LearningRateMultiplierSelection = HyperparameterLearningRateMultiplierSelection.Auto;
options.LearningRateMultiplierSelection = new(1.5); // no clue what real values are!
options.LearningRateMultiplierSelection = 1.5; // again via implicit operator
// This could be something like "EpochCycleCountSelection" or another name with "Epoch" in it if we can make it fit; we just really need the N to turn into a Count somewhere
options.CycleCountSelection = HyperparameterCycleCountSelection.Auto;
options.CycleCountSelection = new(5);
options.CycleCountSelection = 5;
```

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningClient.cs:
```diff
@@ -62,4 +64,50 @@ protected internal FineTuningClient(ClientPipeline pipeline, Uri endpoint, OpenA
+    /// <param name="options"> Additional options (<see cref="RequestOptions"/>) to customize the request. </param>
+    public ClientResult<FineTuningJob> CreateJob(string training_file, string model, RequestOptions options = default)
+    {
+        var request = new InternalCreateFineTuningJobRequest(model, training_file);
```
This "request" type will be need to be converted into a public/visible like like "FineTuningJobCreationOptions" to allow specifying all the optional request settings (like the hyperparameter options, suffix, validation file ID, etc.) -- this is where the signature might be
```
public virtual ClientResult<FineTuningJob> CreateJob(string baseModelName, string trainingFileId, FineTuningJobCreationOptions options = null, CancellationToken cancellationToken = default)
```
It'd otherwise be exact the same, just reusing the options instance if it was already there:
```csharp
options ??= new FineTuningJobCreationOptions();
options.Model = baseModelName;
options.TrainingFile = trainingFileId;
BinaryContent requestContent = options.ToBinaryContent();
...
```

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningClient.cs:
```diff
@@ -62,4 +64,50 @@ protected internal FineTuningClient(ClientPipeline pipeline, Uri endpoint, OpenA
+    /// <param name="training_file"> The training file name that is already uploaded. String should match pattern '^file-[a-zA-Z0-9]{24}$'  </param>
+    /// <param name="model"> The model name to fine-tune. String such as "gpt-3.5-turbo" </param>
+    /// <param name="options"> Additional options (<see cref="RequestOptions"/>) to customize the request. </param>
+    public async Task<ClientResult<FineTuningJob>> CreateJobAsync(string training_file, string model, RequestOptions options = default)
```
hinted at above, but all of these public client methods should include the "virtual" keyword to allow derived overrides

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningClient.cs:
```diff
@@ -62,4 +64,50 @@ protected internal FineTuningClient(ClientPipeline pipeline, Uri endpoint, OpenA
+    /// <returns> A <see cref="Task{T}"/> of <see cref="ClientResult{FineTuningJob}"/> containing the canceled fine-tuning job. </returns>
+    public async Task<ClientResult<FineTuningJob>> CancelJobAsync(string jobId, CancellationToken cancellationToken = default)
+    {
+        ClientResult result = await CancelJobAsync(jobId, cancellationToken.ToRequestOptions());
```
Generally, awaited async method calls inside of libraries always need `.ConfigureAwait(false)` at the end -- separate topic for .NET guts, but it has to do with ensuring thread continuation contexts do the correct/expected things when you're consuming the library (especially if there's any UI interaction involved)

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJob.cs:
```diff
@@ -0,0 +1,23 @@
+[CodeGenModel("FineTuningJob")]
+public partial class FineTuningJob
+{
+    internal static FineTuningJob FromResponse(PipelineResponse pipelineResponse)
```
I think this `FromResponse` should generate in the same form. If it doesn't, let us know and we should fix it!

mikhail replied:
Not sure what you mean here

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJob.cs:
```diff
@@ -0,0 +1,23 @@
+        return DeserializeFineTuningJob(document.RootElement);
+    }
+
+    public FineTuningJobStatus Status { get; }
```
This is just a personal preference thing -- even when I don't change the names, I like to include the explicit `[CodeGenMember("Status")]` attribute; the code generator infers it so long as the name is exactly the same as what it'd generate, but if you renamed later on (e.g. to 'JobStatus', not that it'd make sense here) you'd end up with a painful-to-debug situation where the custom parameter "stopped working" and the old "Status" suspiciously reappeared in generated code. Silent failures are the worst.

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJob.cs:
```diff
@@ -0,0 +1,23 @@
+
+    public FineTuningJobStatus Status { get; }
+
+    public string Object { get; }
```
Can make this private and even a field if desired, e.g. `[CodeGenMember("Object")] private string _object;`

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJob.cs:
```diff
@@ -0,0 +1,23 @@
+    public string Object { get; }
+
+    public FineTuningJobHyperparameters? Hyperparameters { get; set; } 
+    internal IReadOnlyList<InternalFineTuningIntegration> Integrations { get; }
```
When we expose this, we'll probably want to make it a list of "FineTuningIntegrationConfiguration" instances or similar. I know this part is WIP, though!

mikhail replied:
Can't come up with a clean way to handle this. Let's discuss it when there's time. To achieve a simple syntax like this: 

```c
FineTuningJob job = client.CreateJob(
  "gpt-3.5-turbo",
  sampleFile.Id,
  options: new() { Integrations = [new IntegrationWandB("ft-tests")] }
);
```
pretty hacky overrides of getters and setters are needed

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJobHyperparameters.cs:
```diff
@@ -0,0 +1,36 @@
+namespace OpenAI.FineTuning;
+
+[CodeGenModel("FineTuningJobHyperparameters")]
+public partial struct FineTuningJobHyperparameters
```
Mentioned earlier, but more relevant here anyway: this `FineTuningJobHyperparameters` base is the *output* model on the response's `$.hyperparameters` JSON property; you'll see the generated serialization/deserialization code doesn't know anything about batch_size and learning_rate_multiplier since they aren't part of that response-specific schema type.

We'll want two types, I think:
- CreateFineTuningJobRequestHyperparameters -> FineTuningHyperparameterOptions (request model with all three things)
- FineTuningJobHyperparameters -> FineTuningHyperparameterInfo (response model; the "info" might be contentious and we can discuss naming creativity more broadly)

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJobHyperparameters.cs:
```diff
@@ -0,0 +1,36 @@
+
+    // open draft PR to get feedback
+    // re-add hyperparameters
+    [CodeGenMember("n_epochs")]
```
`CodeGenMember` and other attributes will pick up on the processed name rather than the wire one; if you're not customizing the names, it should be identical (e.g. `[CodeGenMember("NEpochs")] public BinaryData NEpochs { get; }` -- if/when you *do* rename, it might not gracefully handle the comparison when there's a not a direct normalized match for "NEpochs" anymore.

trrwilson committed:
initial custom types for request-side ft hyperparameters

Commit:
added test for invalid model

Commit:
separated model name test

mikhail committed:
Grab batch size and LRM from response

mikhail committed:
Fix some tests and fix merge conflicts

mikhail committed:
set NEpochs and BatchSize to be ints while LRM remains float

trrwilson committed:
Merge branch 'main' into user/travisw/custom-ft-types-for-hyperparameter-requests

trrwilson committed:
update: address customization for response hyperparameters

mikhail committed:
Merge remote-tracking branch 'jose/user/travisw/custom-ft-types-for-hyperparameter-requests'

mikhail committed:
Change Hyperparameters yaml definition to include batch and LRM

mikhail committed:
Add WaitUntilCompleted convenience method to FineTuningClient

mikhail committed:
Add PlaygroundURL convenience method to FineTuningJob

mikhail committed:
Add parameters for Suffix and Seed

Introduced _unset_int value to avoid enabling nullable. Arguably not a good approach, but felt like the right thing to do

mikhail committed:
Add WandB integration test for FineTuningClient

mikhail committed:
Allow seed value to be null

Commit:
added validation file id to CreateAndCancelJob and CreateAndCancelJobAsync and tests

Commit:
added CreateAndCancelJobAsync tests

Commit:
Merge branch 'main' of https://github.com/mikhail/openai-in-typespec into main-Mikhail

Commit:
Merged from main

Commit:
handling merge conflicts

mikhail committed:
Reshuffle the tests to be cleaner and more consistent

mikhail committed:
Rename NEpochs to CycleCount and EstimatedFinish+At

mikhail committed:
Rename ValidationFile+Id

mikhail committed:
Rename TrainingFile+Id

mikhail committed:
Rename ResultFiles -> ResultFileIds

mikhail committed:
Update OpenAI.netstandard

mikhail committed:
Convert convenience layers to *Options convention

mikhail committed:
Allow easier comparing of Hyperparams to primitive types

mikhail committed:
Add 'virtual' to method definition in FT Client

mikhail committed:
Add .ConfigureAwait(false) to async calls inside FT Client

mikhail committed:
Improve naming convention

mikhail committed:
Add helper function for Integrations

trrwilson committed:
add spec update for discriminated integration base, user_provided_suffix

trrwilson committed:
FineTuningJobObject

trrwilson committed:
Fix Inner type mapping of wandb (#2)

* add spec update for discriminated integration base, user_provided_suffix

* FineTuningJobObject

* fix inner type mapping of wandb

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix issues not caught during merging

mikhail committed:
Hyperparameters changes to params instead of getters

mikhail committed:
Fix file->sampleFile renaming blunder

mikhail committed:
Remove ConvertTo-Internal

mikhail committed:
Rename model to baseModel

mikhail committed:
Convert GetPlaygroundURL to PlaygroundUri

mikhail committed:
Rename TrainedTokens to BillableTrainedTOkens

mikhail committed:
Rename error param to InvalidParameter

mikhail committed:
Convert InProgress() to parameter and add a test

mikhail committed:
Convert FineTuningJobEvent to public

trrwilson committed:
building test push

mikhail committed:
Fix bugs

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix bugs

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Add FineTuning Events layer with pagination

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Apply dotnet format

mikhail committed:
Use *Options pattern for events

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Address review feedback

mikhail committed:
Rename FineTuningJobError to JobError

mikhail committed:
Remove constructor from HyperparameterOptions

mikhail committed:
Add GetJobs()

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix tests after merge

mikhail committed:
Add a test for GetJobsWithAfter

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Add pagination for GetJobs

mikhail committed:
Remove limit in favor of .Take(). Dedup Sync/Async tests

mikhail committed:
Format all code with dotnet format

mikhail committed:
Convert after and pagesize to ListJobsOptions

mikhail committed:
Add FT Checkpoints code

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix tests after merge

mikhail committed:
Tidy up FineTuningClient.cs

mikhail committed:
Export API

mikhail committed:
Remove cast checking

mikhail committed:
Add test for checkpotin metrics

mikhail committed:
Add assembly reference and regenerate API export

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix tests after merge

mikhail committed:
Add .Value to FTJ Operation

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Tidy up LRO for fine tuning

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
LRO Convenience layer touch up for Fine Tuning

mikhail committed:
Remove debug lines

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJob.cs:
```diff
@@ -0,0 +1,52 @@
+{
+    [CodeGenMember("Id")]
+    public string JobId { get; }
+
```
extra line

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobCheckpoint.cs:
```diff
@@ -0,0 +1,35 @@
+    [CodeGenMember("Object")]
+    private string _object { get; }
+
+    // toString
```
unnecessary comment

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobError.cs:
```diff
@@ -0,0 +1,10 @@
+[CodeGenModel("FineTuningJobError")]
+public partial class JobError
+{
+
```
extra line

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobError.cs:
```diff
@@ -0,0 +1,10 @@
+
+    [CodeGenMember("Param")]
+    public string InvalidParameter { get; }
+
```
extra line

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobHyperparameters.cs:
```diff
@@ -0,0 +1,46 @@
+using System.Collections.Generic;
+
+namespace OpenAI.FineTuning;
+
```
extra line

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobOperation.Protocol.cs:
```diff
@@ -14,25 +15,33 @@ namespace OpenAI.FineTuning;
 
-    private readonly string _jobId;
+
+    new public bool HasCompleted
```
I might have missed this in the chat, but does this mean the user needs to call UpdateStatus to get a current bool value here? Would it be better to make a call here if it hasn't been updated in a while?

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobOperation.Protocol.cs:
```diff
@@ -14,25 +15,33 @@ namespace OpenAI.FineTuning;
 {
-    private readonly ClientPipeline _pipeline;
-    private readonly Uri _endpoint;
+    internal readonly ClientPipeline _pipeline;
```
do these get used somewhere outside of this class? Why change to internal?

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobOperation.Protocol.cs:
```diff
@@ -44,122 +53,84 @@ internal FineTuningJobOperation(
+    /// <returns> The rehydrated operation <see cref="FineTuningJobOperation"/>. </returns>
     /// <exception cref="ArgumentNullException"> <paramref name="client"/> or <paramref name="rehydrationToken"/> is null. </exception>
-    public static async Task<FineTuningJobOperation> RehydrateAsync(FineTuningClient client, ContinuationToken rehydrationToken, CancellationToken cancellationToken = default)
+    public static FineTuningJobOperation Rehydrate(FineTuningClient client, ContinuationToken rehydrationToken, RequestOptions options)
```
I think it might be better to keep this as a specified CancellationToken so people know that is what it is used for. Theoretically the actual request options if any are kept in the ContinuationToken right?

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobOperation.Protocol.cs:
```diff
@@ -44,122 +53,84 @@ internal FineTuningJobOperation(
-    /// <exception cref="ArgumentNullException"> <paramref name="client"/> or <paramref name="rehydrationToken"/> is null. </exception>
-    public static FineTuningJobOperation Rehydrate(FineTuningClient client, ContinuationToken rehydrationToken, CancellationToken cancellationToken = default)
+    /// <inheritdoc cref="Rehydrate(FineTuningClient, ContinuationToken, RequestOptions)"/>
+    public static async Task<FineTuningJobOperation> RehydrateAsync(FineTuningClient client, ContinuationToken rehydrationToken, RequestOptions options)
```
I think it might be better to keep this as a specified CancellationToken so people know that is what it is used for. Theoretically the actual request options if any are kept in the ContinuationToken right?

mikhail replied:
I think CancellationToken shouldn't be used in Protocol layer. There's a copy of this method in the convenience layer with cancellation token

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobOperation.cs:
```diff
@@ -0,0 +1,206 @@
+    /// <param name="cancellationToken"> The cancellation token to use. </param>"
+    /// <exception cref="ClientResultException"> Service returned a non-success status code. </exception>
+    /// <returns> The response returned from the service. </returns>
+    public virtual async Task<ClientResult<FineTuningJob>> GetJobAsync(CancellationToken cancellationToken = default)
```
should this just call the internal one? Same with the sync

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningJobOperation.cs:
```diff
@@ -0,0 +1,206 @@
+        return NewMethod(method, out message, cancellationToken);
+    }
+
+    private ClientResult<FineTuningJob> NewMethod(Func<string, RequestOptions?, PipelineMessage> method, out PipelineMessage message, CancellationToken cancellationToken)
```
What does this do? Maybe a comment to describe it?

mikhail replied:
Whoa, looks like an unintentional refactor. Removed.

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningOptions.cs:
```diff
@@ -0,0 +1,64 @@
+[CodeGenModel("CreateFineTuningJobRequest")]
+public partial class FineTuningOptions
+{
+
```
extra line

chschrae commented on .dotnet/src/Custom/FineTuning/FineTuningOptions.cs:
```diff
@@ -0,0 +1,64 @@
+    /// </summary>
+    [CodeGenMember("Integrations")]
+    public IList<FineTuningIntegration> Integrations { get; }
+
```
extra line

mikhail commented on .dotnet/src/Custom/FineTuning/FineTuningClient.Protocol.cs:
```diff
@@ -53,9 +56,8 @@ public virtual async Task<FineTuningJobOperation> CreateFineTuningJobAsync(
         string jobId = doc.RootElement.GetProperty("id"u8).GetString();
-        string status = doc.RootElement.GetProperty("status"u8).GetString();
 
-        FineTuningJobOperation operation = this.CreateCreateJobOperation(jobId, status, response);
```
Status is already inside of response. no need to duplicate. I was considering getting rid of jobId as well. It makes sense that to recreate a JobOperation all you need is the response

mikhail committed:
Call Protocol method from convenience layer when possible.

mikhail committed:
CRLF -> LF

mikhail commented on .dotnet/src/Custom/FineTuning/FineTuningClient.cs:
```diff
@@ -1,90 +1,176 @@
-        return new FineTuningJobOperation(_pipeline, _endpoint, jobId, status, response);
-    }
-}
+using System;
```
is LF what it should be?

mikhail committed:
PR comments. Internal/Private. Whitespace.

mikhail committed:
Set HasCompleted on result update. Fix GetJob casting

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Make FineTuningOperation primary class and hide 'Job'

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Rerun codegen

mikhail committed:
Improve naming around PipelineMessages

mikhail committed:
Improve method naming around FineTuning

mikhail committed:
Improve class naming around FineTuning

mikhail committed:
Rename FineTuningJobId to JobId

mikhail committed:
Rename all Operation to Job

mikhail committed:
Final touches renaming Operation to Job

mikhail committed:
Minor renames in preliminary APIView review

mikhail committed:
Minor renames in preliminary APIView review, part 2

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix Azure FT tests

mikhail committed:
Fix Azure FT file name typo

mikhail committed:
Merge remote-tracking branch 'jose/main' compilation is failing

mikhail committed:
Fixing post merge issues

mikhail committed:
Fixing post merge issues pt 2

trrwilson committed:
Partial port of spec update for "method" + hyperparameter contemporaries (#3)

* initial rough commit, request models represented (but not response)

* port latest spec changes; compilation works

* CodeGenModel cleanup

* Fix method type and equality tests

---------

Co-authored-by: Mikhail Simin <mikhail.simin@gmail.com>

mikhail committed:
Debug CheckpointsFineTuning()

mikhail committed:
Remove API Version from pipeline message builder

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Fix up smoke tests

mikhail committed:
Mark deletion test as Live

mikhail committed:
Use the new assets tag

mikhail committed:
Use the new assets tag and comment out ConversationTests rate limits

mikhail committed:
Bump assets tag

mikhail committed:
Mark everything as experimental

mikhail committed:
Update OpenAI.netstandard

trrwilson commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
     }
 }
 namespace OpenAI.FineTuning {
+    public readonly partial struct CreateFineTuningJobRequestModel : IEquatable<CreateFineTuningJobRequestModel> {
```
Thus far, I believe we've kept these enumerated types for model choices internal and done representation as plain `string`, with better mechanisms for known model selection an open architecture/design problem to solve. Even if just for consistency, can we do the same here and make these model name extensible enums internal (e.g. via "generatorstubs") and swap their references to strings?

mikhail replied:
luckily no references

trrwilson commented on .dotnet/src/Custom/FineTuning/FineTuningJob.cs:
```diff
@@ -0,0 +1,272 @@
+
+    internal FineTuningHyperparameters Hyperparameters { get; private set; } = default;
+    public IReadOnlyList<FineTuningIntegration> Integrations { get; private set; } = null!;
+    public int? BillableTrainedTokens { get; private set; }
```
Minor thing for consistency: make this a non-nullable `int` (will default to 0 when not present, which is accurate) and name it `BillableTrainedTokenCount` (see: ChatCompletionUsage, etc.).

I'm crying a little inside that this is the one place where OpenAI apparently *doesn't* have a dedicated `usage` container with expandable breakouts for details. We might consider adding that ourselves, but not for this PR.

mikhail replied:
I'd have to change it in the TypeSpec as well. That's ok? It explicitly calls out that the value is null while the job is still running

mikhail replied:
![image](https://github.com/user-attachments/assets/17d23958-66a2-4514-83f0-68892307e4d0)

mikhail committed:
Convert RequestModel to String

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+    public class FineTuningCheckpoint : IJsonModel<FineTuningCheckpoint>, IPersistableModel<FineTuningCheckpoint> {
+        public string CheckpointId { get; }
+        public DateTimeOffset CreatedAt { get; }
+        public string FineTunedModelCheckpointId { get; }
```
The way OpenAI named this property is a little weird. It is not a "checkpoint ID" but a "model ID". More specifically, it's the ID of the model that was created for this checkpoint. I wonder if we can rename this to make it clearer. Perhaps we can call it `CheckpointModelId`?

mikhail replied:
Yes that's what it means. Similar to comment below - could it be as simple as `ModelId` ? so referencing it would be `myCheckpoint.ModelId`

joseharriaga replied:
Yes, that makes sense!

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public override readonly string ToString();
+    }
+    public class FineTuningCheckpoint : IJsonModel<FineTuningCheckpoint>, IPersistableModel<FineTuningCheckpoint> {
+        public string CheckpointId { get; }
```
We can call this simply `Id`. Since the class is called `FineTuningCheckpoint`, it is clear that it is the checkpoint ID.

mikhail replied:
Agreed. I think I misunderstood a conversation in a meeting instructing me to do the opposite. Will rename.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public float? TrainLoss { get; }
+        public float? TrainMeanTokenAccuracy { get; }
+        public float? ValidLoss { get; }
+        public float? ValidMeanTokenAccuracy { get; }
```
Just double-checking: Are these six properties all optional? I haven't tried it personally, but every example I've seen online has non-null values for all six of them. Not sure if it's a coincidence, or if these are not actually optional?

mikhail replied:
Will check in on this.

mikhail replied:
OpenAI's spec doesn't list them as required. Similarly to Event.data comment - it's a reasonable guess that they will be there but a guess nonetheless

joseharriaga replied:
Sounds good, we can keep them optional for now.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public virtual FineTuningJob FineTune(BinaryContent content, bool waitUntilCompleted, RequestOptions options);
+        public virtual FineTuningJob FineTune(string baseModel, string trainingFileId, FineTuningOptions options = null, CancellationToken cancellationToken = default);
+        public virtual Task<FineTuningJob> FineTuneAsync(BinaryContent content, bool waitUntilCompleted, RequestOptions options);
+        public virtual Task<FineTuningJob> FineTuneAsync(string baseModel, string trainingFileId, FineTuningOptions options = null, CancellationToken cancellationToken = default);
```
All four `FineTune` methods should take a `bool waitUntilCompleted` parameter.

mikhail replied:
done

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public FineTuningJob GetJob(string JobId, CancellationToken cancellationToken = default);
+        public Task<FineTuningJob> GetJobAsync(string JobId, CancellationToken cancellationToken = default);
+        public virtual CollectionResult<FineTuningJob> ListJobs(ListJobsOptions options = null, CancellationToken cancellationToken = default);
+        public virtual AsyncCollectionResult<FineTuningJob> ListJobsAsync(ListJobsOptions options = null, CancellationToken cancellationToken = default);
```
Rather than `ListJobs`, in .NET we prefer `GetJobs`. Similarly, instead of `ListJobsOptions`, we have been calling them `FineTuningJobCollectionOptions`.

mikhail replied:
Done

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public static implicit operator BinaryContent(FineTuningError fineTuningError);
+    }
+    public class FineTuningEvent : IJsonModel<FineTuningEvent>, IPersistableModel<FineTuningEvent> {
+        public string Level;
```
1. According to the spec, this could be an enum with values: "info", "warn", or "error".

2. This property represents a "log level", so I recommend renaming it to `LogLevel`.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public BinaryData Data { get; }
+        public string Id { get; }
+        public string Message { get; }
+        public FineTuningJobEventType? Type { get; }
```
We prefer using `Kind` instead of `Type`, because the word "type" has a special connotation in .NET.

mikhail replied:
Currently in this repo there are 12 instances of `Kind` and 46 instances of `Type`. Are you sure we want to weigh in on Kind? Happy to do it; just pointing out the divergence

joseharriaga replied:
Yes, `Kind` is the right one. We need to go and fix those instances of `Type`. I believe they should all be in Experimental APIs, so we can make breaking changes to fix them, but would be good to avoid introducing more.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+    public class FineTuningEvent : IJsonModel<FineTuningEvent>, IPersistableModel<FineTuningEvent> {
+        public string Level;
+        public DateTimeOffset CreatedAt { get; }
+        public BinaryData Data { get; }
```
I wonder if there is a reasonable way to be able to deserialize this data into something usable instead of using `BinaryData`. Looking at an example, an event object looks like this:

```json
{
    "object": "fine_tuning.job.event",
    "id": "ftevent-abc-123",
    "created_at": 1693582679,
    "level": "info",
    "message": "Step 300/300: training loss=0.15, validation loss=0.27, full validation loss=0.40",
    "data": {
        "step": 300,
        "train_loss": 0.14991648495197296,
        "valid_loss": 0.26569826706596045,
        "total_steps": 300,
        "full_valid_loss": 0.4032616495084362,
        "train_mean_token_accuracy": 0.9444444179534912,
        "valid_mean_token_accuracy": 0.9565217391304348,
        "full_valid_mean_token_accuracy": 0.9089635854341737
    },
    "type": "metrics"
}
```

Should we create a `FineTuningJobEventData` class with those properties and use that instead?

mikhail replied:
`data` is not defined in the OpenAPI spec (yaml). I could make a reasonable guess, but it would be a guess from me rather than from OpenAI's side.

joseharriaga replied:
We can leave it as `BinaryData` for now.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+    }
+    public readonly partial struct FineTuningHyperparameters : IJsonModel<FineTuningHyperparameters>, IPersistableModel<FineTuningHyperparameters>, IJsonModel<object>, IPersistableModel<object> {
+        public int BatchSize { get; }
+        public int CycleCount { get; }
```
As enigmatic as it sounds at first, I think the industry term for this is "epoch", so we should probably name it `EpochCount`.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public static explicit operator FineTuningEvent(ClientResult result);
+        public static implicit operator BinaryContent(FineTuningEvent fineTuningEvent);
+    }
+    public readonly partial struct FineTuningHyperparameters : IJsonModel<FineTuningHyperparameters>, IPersistableModel<FineTuningHyperparameters>, IJsonModel<object>, IPersistableModel<object> {
```
1. Should this be a class rather than a struct?
2. Is this type used anywhere?

mikhail replied:
It's still used, although marked Obsolete as OpenAI has deprecated the property. Struct made some things a lot easier, so I made that call. Won't matter soon anyway

mikhail replied:
Let's discuss whether it makes sense to release something brand new that's marked as deprecated.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1743,37 +1743,291 @@ public static class OpenAIFilesModelFactory {
+        public static explicit operator FineTuningHyperparameters(ClientResult result);
+        public static implicit operator BinaryContent(FineTuningHyperparameters fineTuningHyperparameters);
+    }
+    public abstract class FineTuningIntegration : IJsonModel<FineTuningIntegration>, IPersistableModel<FineTuningIntegration> {
```
This should have a factory method that allows you to create a `WeightsAndBalancesFineTuningIntegration`. Something like this:

```csharp
public static WeightAndBalancesFineTuningIntegration CreateWeightsAndBalancesIntegration(string projectName);
```

mikhail replied:
class WeightsAndBiasesIntegration is public and inherits from FineTuningIntegration. Do we still need a static factory?

mikhail committed:
Introduce MethodHyperparameters

mikhail committed:
Merge remote-tracking branch 'jose/main'

joseharriaga commented:
I made a few improvements to the Export-Api.ps1 script. It now produces a file called OpenAI.net8.0.cs which is the same as OpenAI.netstandard2.0.cs with the difference that this one does show the `[Experimental]` attribute so we can actually verify that everything in the FineTuning namespace is marked as experimental.

mikhail committed:
Get to compile after merge

mikhail committed:
AOAI FineTuning missing version in api call

mikhail committed:
Revert "Remove API Version from pipeline message builder"

This reverts commit 906bce87b59e6580975e7bc52231c98911686d13.

joseharriaga commented on .dotnet/api/OpenAI.netstandard2.0.cs:
```diff
@@ -1763,37 +1763,291 @@ public static class OpenAIFilesModelFactory {
     }
 }
 namespace OpenAI.FineTuning {
```
I made a few improvements to the Export-Api.ps1 script. It now produces a file called OpenAI.net8.0.cs which is the same as OpenAI.netstandard2.0.cs with the difference that this one does show the [Experimental] attribute so we can actually verify that everything in the FineTuning namespace is marked as experimental.

mikhail replied:
Regenerated

mikhail committed:
Broken: getting to azureFineTuningJob

mikhail committed:
CreateCancelDelete test works

mikhail committed:
Fix tests after merge

mikhail committed:
Allow Running status for intentinally bad files

mikhail committed:
Merge remote-tracking branch 'jose/main'

mikhail committed:
Remove Record test mode

mikhail committed:
Mark Deploy test as Explicit and bump assets

mikhail committed:
Address naming change requests and regenerate API

mikhail committed:
Merge remote-tracking branch 'jose/main'
